File: test_arithmetic_coding.py
import math, random
from arithmetic_coding_alt import ArithmeticCoder

# from arithmetic_coding import ArithmeticCoder


def get_frequencies(text):
    "Estimate frequency table from given text."
    freqs = dict()
    for c in text:
        freqs[c] = freqs.get(c, 0) + 1

    return freqs


def test(message, frequencies=None):
    if frequencies is None:
        frequencies = get_frequencies(message)

    endec = ArithmeticCoder(frequencies, mode="dc", register_size=16)
    encoded = endec.encode(message)
    decoded = endec.decode(encoded, len(message))

    # encode and decode message
    encoded_string = "".join(str(bit) for bit in encoded)
    entropy = len(message) * math.log2(sum(frequencies.values())) - sum(
        math.log2(frequencies[c]) for c in message
    )

    # print info
    maxlen = 128
    print(f"Entropy of msg = {entropy:.2f} bits")
    print(f"Encoded length = {len(encoded)} bits")

    if len(encoded) < maxlen:
        print(f'Message string = "{message}"')
    else:
        print(f'Message string = "{message[:maxlen]}..."')

    if len(encoded_string) < maxlen:
        print(f"Encoded string = {encoded_string}")
    else:
        print(f"Encoded string = {encoded_string[:maxlen]}...")

    if len(decoded) < maxlen:
        print(f'Decoded string = "{decoded}"')
    else:
        print(f'Decoded string = "{decoded[:maxlen]}..."')

    # check if decoding was successful
    assert message == decoded


# test 1
print("\nTest 1\n======")
message = "ABBY CADABBY"
frequencies = {"A": 126, "B": 167, "C": 116, "D": 88, "Y": 89, " ": 100}
test(message, frequencies)

# test2
print("\nTest 2\n======")
with open("test_arithmetic_coding.py") as f:
    message = f.read()
test(message)

# test 3 - compress random strings
print("\nTest 3\n======")
dc = ArithmeticCoder(frequencies, mode="dc", register_size=16, pad=1)

letters = list(frequencies.keys())
weights = list(frequencies.values())
for trial in range(10000):
    x = "".join(random.choices(letters, weights=weights, k=10))
    y = dc.encode(x)
    xh = dc.decode(y, len(x))

    print(f"\nTrial: {trial}")
    print(x)
    print("".join(str(i) for i in y))
    print(xh)

    assert x == xh

# test 4 - generate random strings
print("\nTest 4\n======")
sg = ArithmeticCoder(frequencies, mode="sg", register_size=16, pad=1)

for trial in range(10000):
    x = [random.randint(0, 1) for _ in range(30)]
    y = sg.encode(x)
    xh = sg.decode(y, len(x))

    print(f"\nTrial: {trial}")
    print("".join(str(i) for i in x))
    print(y)
    print("".join(str(i) for i in xh))

    assert x == xh

File: test_arithmetic_coding.jl
using StatsBase

include("arithmetic_coding_alt.jl")
# include("arithmetic_coding.jl")

function get_frequencies(text::String)
    freqs = Dict{Char, Int}()
    for c in text
        freqs[c] = get(freqs, c, 0) + 1
    end

    return freqs
end

function test(message, frequencies = nothing)
    if frequencies == nothing
        frequencies = get_frequencies(message)
    end
    endec = make_endec(frequencies, "dc")
    encoded = encode(endec, message)
    decoded = decode(endec, encoded, length(message))

    # encode and decode message
    encoded_string = join(encoded)
    entropy = length(message) * log2(sum(values(frequencies))) - sum(log2(frequencies[c]) for c in message)
    entropy = round(entropy, digits = 2)

    # print info
    println("Entropy of msg = $(entropy) bits")
    println("Encoded length = $(length(encoded)) bits")

    maxlen = 128
    if length(encoded) < maxlen
        println("Message string = \"$message\"")
    else
        println("Message string = \"$(message[1:maxlen])...\"")
    end

    if length(encoded_string) < maxlen
        println("Encoded string = $encoded_string")
    else
        println("Encoded string = $(encoded_string[1:maxlen])...")
    end

    if length(decoded) < maxlen
        println("Decoded string = \"$decoded\"")
    else
        println("Decoded string = \"$(decoded[1:maxlen])...\"")
    end

    @assert message == decoded
end

# test 1
println("\nTest 1\n======")
message = "ABBY CADABBY"
frequencies = Dict('A' => 126, 'B' => 167, 'C' => 116, 'D' => 88, 'Y' => 89, ' ' => 100)
test(message, frequencies)

# test 2
println("\nTest 2\n======")
message = read("test_arithmetic_coding.jl", String)
test(message)

# test 3 - compress random strings
println("\nTest 3\n======")
dc = make_endec(frequencies, "dc", register_size = 16, pad = 1)

letters = collect(keys(frequencies))
weights = collect(values(frequencies))

for trial in 1:10000
    x = join(sample(letters, ProbabilityWeights(weights), 10))
    y = encode(dc, x)
    xh = decode(dc, y, length(x))

    println("\nTrial: $(trial)")
    println(x)
    println(join(y))
    println(xh)

    @assert x == xh
end

# test 4 - generate random strings
println("\nTest 4\n======")
sg = make_endec(frequencies, "sg", register_size = 16, pad = 1)
for trial in 1:10000
    x = rand(0:1, 30)
    y = encode(sg, x)
    xh = decode(sg, y, length(x))

    println("\nTrial: $(trial)")
    println(join(x))
    println(y)
    println(join(xh))

    @assert x == xh
end
